// @milanzor/homey-mock/includes/Device/Device.js
/**
 *
 */
class Device {

    /**
     *
     * @returns {Device}
     */
    getAvailable() {
        return this;
    }

    /**
     *
     * @returns {Device}
     */
    getCapabilities() {
        return this;
    }

    /**
     *
     * @returns {Device}
     */
    getCapabilityValue() {
        return this;
    }

    /**
     *
     * @returns {Device}
     */
    getClass() {
        return this;
    }

    /**
     *
     * @returns {Device}
     */
    getData() {
        return this;
    }

    /**
     *
     * @returns {Device}
     */
    getDriver() {
        return this;
    }

    /**
     *
     * @returns {Device}
     */
    getName() {
        return this;
    }

    /**
     *
     * @returns {Device}
     */
    getSetting() {
        return this;
    }

    /**
     *
     * @returns {Device}
     */
    getSettings() {
        return this;
    }

    /**
     *
     * @returns {Device}
     */
    getState() {
        return this;
    }

    /**
     *
     * @returns {Device}
     */
    getStore() {
        return this;
    }

    /**
     *
     * @returns {Device}
     */
    getStoreKeys() {
        return this;
    }

    /**
     *
     * @returns {Device}
     */
    getStoreValue() {
        return this;
    }

    /**
     *
     * @returns {Device}
     */
    hasCapability() {
        return this;
    }

    /**
     *
     * @returns {Device}
     */
    onAdded() {
        return this;
    }

    /**
     *
     * @returns {Device}
     */
    onDeleted() {
        return this;
    }

    /**
     *
     * @returns {Device}
     */
    onInit() {
        return this;
    }

    /**
     *
     * @returns {Device}
     */
    onRenamed() {
        return this;
    }

    /**
     *
     * @returns {Device}
     */
    onSettings() {
        return this;
    }

    /**
     *
     * @returns {Device}
     */
    ready() {
        return this;
    }

    /**
     *
     * @returns {Device}
     */
    registerCapabilityListener() {
        return this;
    }

    /**
     *
     * @returns {Device}
     */
    registerMultipleCapabilityListener() {
        return this;
    }

    /**
     *
     * @returns {Device}
     */
    setAlbumArtImage() {
        return this;
    }

    /**
     *
     * @returns {Device}
     */
    setAvailable() {
        return this;
    }

    /**
     *
     * @returns {Device}
     */
    setCapabilityValue() {
        return this;
    }

    /**
     *
     * @returns {Device}
     */
    setSettings() {
        return this;
    }

    /**
     *
     * @returns {Device}
     */
    setStoreValue() {
        return this;
    }

    /**
     *
     * @returns {Device}
     */
    setUnavailable() {
        return this;
    }

    /**
     *
     * @returns {Device}
     */
    setWarning() {
        return this;
    }

    /**
     *
     * @returns {Device}
     */
    triggerCapabilityListener() {
        return this;
    }

    /**
     *
     * @returns {Device}
     */
    unsetStoreValue() {
        return this;
    }

    /**
     *
     * @returns {Device}
     */
    unsetWarning() {
        return this;
    }
}

module.exports = Device;
