// @milanzor/homey-mock/includes/Managers/ManagerApi.js
/**
 *
 */
class ManagerApi {

    /**
     *
     * @returns {ManagerApi}
     */
    delete() {
        return this;
    }

    /**
     *
     * @returns {ManagerApi}
     */
    get() {
        return this;
    }

    /**
     *
     * @returns {ManagerApi}
     */
    getApi() {
        return this;
    }

    /**
     *
     * @returns {ManagerApi}
     */
    getLocalUrl() {
        return this;
    }

    /**
     *
     * @returns {ManagerApi}
     */
    getOwnerApiToken() {
        return this;
    }

    /**
     *
     * @returns {ManagerApi}
     */
    post() {
        return this;
    }

    /**
     *
     * @returns {ManagerApi}
     */
    put() {
        return this;
    }

    /**
     *
     * @returns {ManagerApi}
     */
    realtime() {
        return this;
    }

    /**
     *
     * @returns {ManagerApi}
     */
    registerApi() {
        return this;
    }

    /**
     *
     * @returns {ManagerApi}
     */
    unregisterApi() {
        return this;
    }
}

module.exports = ManagerApi;
