// @milanzor/homey-mock/includes/Ledring/LedringAnimationSystemProgress.js
/**
 *
 */
class LedringAnimationSystemProgress {

    /**
     *
     * @returns {LedringAnimationSystemProgress}
     */
    register() {
        return this;
    }

    /**
     *
     * @returns {LedringAnimationSystemProgress}
     */
    registerScreensaver() {
        return this;
    }

    /**
     *
     * @returns {LedringAnimationSystemProgress}
     */
    setProgress() {
        return this;
    }

    /**
     *
     * @returns {LedringAnimationSystemProgress}
     */
    start() {
        return this;
    }

    /**
     *
     * @returns {LedringAnimationSystemProgress}
     */
    stop() {
        return this;
    }

    /**
     *
     * @returns {LedringAnimationSystemProgress}
     */
    unregister() {
        return this;
    }

    /**
     *
     * @returns {LedringAnimationSystemProgress}
     */
    unregisterScreensaver() {
        return this;
    }

    /**
     *
     * @returns {LedringAnimationSystemProgress}
     */
    updateFrames() {
        return this;
    }
}

module.exports = LedringAnimationSystemProgress;
