// @milanzor/homey-mock/includes/Ble/BlePeripheral.js
/**
 *
 */
class BlePeripheral {
    
    /**
     *
     * @returns {BlePeripheral}
     */
    assertConnected() {
        return this;
    }

    /**
     *
     * @returns {BlePeripheral}
     */
    connect() {
        return this;
    }

    /**
     *
     * @returns {BlePeripheral}
     */
    disconnect() {
        return this;
    }

    /**
     *
     * @returns {BlePeripheral}
     */
    discoverAllServicesAndCharacteristics() {
        return this;
    }

    /**
     *
     * @returns {BlePeripheral}
     */
    discoverServices() {
        return this;
    }

    /**
     *
     * @returns {BlePeripheral}
     */
    getService() {
        return this;
    }

    /**
     *
     * @returns {BlePeripheral}
     */
    read() {
        return this;
    }

    /**
     *
     * @returns {BlePeripheral}
     */
    updateRssi() {
        return this;
    }

    /**
     *
     * @returns {BlePeripheral}
     */
    write() {
        return this;
    }

    /**
     *
     * @returns {BlePeripheral}
     */
    readValue() {
        return this;
    }

    /**
     *
     * @returns {BlePeripheral}
     */
    writeValue() {
        return this;
    }

}

module.exports = BlePeripheral;
